src/RecordData/DataFormatter.cpp

```diff
@@ -0,0 +1,127 @@
+    offset += state->getPackedDataSize();
+
+    // pack sensor data
+    for (int i = 0; i < state->getNumMaxSensors(); i++)
+    {
+        if (!state->sensorOK(state->getSensors()[i]))
+            continue;
+        memcpy((uint8_t *)dest + offset, state->getSensors()[i]->getPackedData(), state->getSensors()[i]->getPackedDataSize());
+        offset += state->getSensors()[i]->getPackedDataSize();
+    }
+
+    return dest + offset;
+}
+
+int DataFormatter::getPackedLen(State *state)
+{
+    int len = state->getPackedDataSize();
+    for (int i = 0; i < state->getNumMaxSensors(); i++)
+    {
+        if (!state->sensorOK(state->getSensors()[i]))
+            continue;
+        len += state->getSensors()[i]->getPackedDataSize();
+    }
+    return len;
+}
+
+void DataFormatter::getCSVHeader(char *dest, int destLen, State *state)
+{
+    int offset = 0;
+    // state first
+    const char **labels = state->getPackedDataLabels();
+    for (int i = 0; i < state->getNumPackedDataPoints() && destLen > 0; i++)
+    {
+        int len = strlen(labels[i]) + 1;
+        snprintf(dest + offset, destLen, "%s,", labels[i]);
+        offset += len;
+        destLen -= len;
+    }
+
+    // sensors next
+    for (int i = 0; i < state->getNumMaxSensors(); i++)
+    {
+        if (!state->sensorOK(state->getSensors()[i]))
+            continue;
+        labels = state->getSensors()[i]->getPackedDataLabels();
+        for (int j = 0; j < state->getSensors()[i]->getNumPackedDataPoints() && destLen > 0; j++)
+        {
+
+            int len = snprintf(dest + offset, destLen, "%s,", labels[j]);
+            offset += len;
+            destLen -= len;
+        }
+    }
+    dest[offset - 1] = '\0';
+}
+
+void DataFormatter::toCSVRow(char *dest, int destLen, State *state, void *data)
+{
+    int dataOffset = 0;
+    // state first
+    uintptr_t cursor = toCSVSection(dest, destLen, data == nullptr ? state->getPackedData() : (void *)data, dataOffset, state);
+
+    // sensors next
+    for (int i = 0; i < state->getNumMaxSensors(); i++)
+    {
+        if (!state->sensorOK(state->getSensors()[i]))
+            continue;
+        if (data == nullptr)
+            dataOffset = 0;
+        cursor = toCSVSection((char *)cursor, destLen, data == nullptr ? state->getSensors()[i]->getPackedData() : (void *)data, dataOffset, state->getSensors()[i]);
+    }
+    ((char *)cursor)[-1] = '\0';
+}
+
+uintptr_t DataFormatter::toCSVSection(char *dest, int &destLen, void *data, int &dataOffset, DataReporter *obj)
+{
+    int strSize = 0;
+    for (int i = 0; i < obj->getNumPackedDataPoints() && destLen > 0; i++)
+    {
+        const void *dataPtr = (uint8_t *)data + dataOffset;
+        dataOffset += obj->PackedTypeToSize(obj->getPackedOrder()[i]);
+        int printedSize = 0;
+        switch (obj->getPackedOrder()[i])
+        {
+        case FLOAT:
+            printedSize = snprintf(dest + strSize, destLen, "%.3f,", *(float *)dataPtr);
```

Comment:
Can you use the return value like this?

"*The snprintf() function returns the number of characters that would have been written had n been sufficiently large, not counting the terminating null character, or a negative value if an encoding error occurred*"

Reply:
yes it's the number readable of characters printed. If destLen runs out of space, the next iteration of the for loop will catch that and stop any further print attempts